### PR TITLE
Fix GC: don't mark pointer outside of megablock refs #48

### LINKED ISSUE #48: Segmentation fault: gc_add_root with static thunk
When static thunk is added to gc root, `gc_perform_gc` is crashed by segmentation fault.
### Detail
- static thunk is store its value in it.
- `gc_mark_deeper` peeks the value, and regards it as pointer.
- `gc_add_grey` regards it is in heap, and set used bits
- crash!!
### Example code

Because Ajhc's optimaization is so powerful, static thunk is rare generated. So, we write C code directly.

``` c
#include "jhc_rts_header.h"

struct my_entry {  int n; };

int main(int argc, char *argv[])
{
  gc_t gc;
  arena_t arena;
  jhc_alloc_init(&gc, &arena);
  struct s_cache *sc = new_cache(arena,sizeof(struct my_entry),1);
  wptr_t ptr = s_alloc(gc, arena, sc);
  ((struct my_entry*)ptr)->n = 0x200; // pointer like value
  gc_add_root(gc, arena, ptr);

  gc_perform_gc(gc, arena);
  return 0;
}
```
### How to build

Put it in 'rts/test/' and build:

```
gcc -I.. -I../.. -std=gnu99 -g -O2 -D_JHC_GC=_JHC_GC_JGC  -DJHC_UNIT -D_JHC_STANDALONE=0 -DJHC_VALGRIND=1    gc_root_test.c hs_fake.c ../rts/profile.c ../rts/jhc_rts.c ../rts/gc_jgc.c ../rts/stableptr.c ../rts/gc_none.c ../rts/rts_support.c ../rts/conc.c   -o gc_root_test
```

### PR DESCRIPTION
**Pull request by mzp**
Fix #48.

### CODE REVIEW

**Comment by master-q** [comment]
Umm.... This patch is too slow.
I am thinking better code...

**Comment by mzp** [comment]
Oh.., I agree with your opinion. But I have no idea for general case.
- When `_JHC_JGC_LIMITED_NUM_MEGABLOCK`  is enable, we could use more smart condition
- If addresses of heap memory is consecutive, we could use more smart condition

File: rts/test/gc_root_test.c [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+int
+main(int argc, char *argv[])
+{
+  gc_t gc;
+  arena_t arena;
+  jhc_alloc_init(&gc, &arena);
+  struct s_cache *sc = new_cache(arena,sizeof(struct my_entry),1);
+  wptr_t ptr = s_alloc(gc, arena, sc);
+  ((struct my_entry*)ptr)->n = 0x200;
```
Comment (master-q) [inline rts/test/gc_root_test.c:16]: I think jgc interprets 0x200 as strict pointer.
Then jgc will track the smart pointer with gc_add_grey(), and crash.
I think it's correct.

**Comment by mzp** [comment]
Oops, this patch does not fix #48.

**Pull request closed without merging**